src/System.Management.Automation/engine/runtime/Binding/Binders.cs

```diff
@@ -2029,36 +2029,34 @@ private static object NullRule(CallSite site, object obj)
 
         internal static bool IsValueTypeMutable(Type type)
         {
-            if (type.IsPrimitive || type.IsEnum)
+            // First, check compiler-defined attributes. This could be the builtin type, or a shoe-in by the compiler
+            // if the library targets netstandard2.0, where the compiler can't be sure the attribute will be provided
+            // by the runtime, and defines its own attribute of the same name during compilation.
+            if (type.IsDefined(typeof(System.Runtime.CompilerServices.IsReadOnlyAttribute), inherit: false))
             {
                 return false;
             }
 
-            // If there are any fields, the type is mutable.
-            if (type.GetFields(BindingFlags.Public | BindingFlags.Instance).Any())
+            foreach (var attribute in type.GetCustomAttributes(inherit: false))
             {
-                return true;
+                if (attribute.GetType().FullName.Equals(
+                    "System.Runtime.CompilerServices.IsReadOnlyAttribute",
+                    StringComparison.Ordinal))
+                {
+                    return false;
+                }
             }
 
-            // If there are any properties with setters, the type is mutable.
-            var properties = type.GetProperties(BindingFlags.Public | BindingFlags.Instance);
-            for (int index = 0; index < properties.Length; index++)
+            // Fallback: check all fields (public + private) to verify whether they're all readonly.
+            // If any field is not readonly, the value type is mutable.
+            foreach (var field in type.GetFields(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic))
             {
-                var property = properties[index];
-                if (property.CanWrite)
+                if (!field.IsInitOnly)
```

Comment:
What is a private field is not readonly but a property (used the field) is?
I think we should revert the code and enhance a code for methods below to take into account new readonly methods.

Reply:
The private field can still mutate, e.g., if the type provides a method that modifies the private field. Without checking all the fields, we cannot be sure the struct is immutable.

Also, even if we could check it more heuristically like you suggest, it's likely that doing more complex logic would be more expensive, since we'd need to make several different checks and verify a wider scope of things than just checking all the fields.

Reply:
Please re-read my _counterexample_:
> What is a private field is not readonly but a property (used the field) is?
```c#
using System;
struct S
{
    private int a;
    S(int p)
    {
        a = 1;
    }
    public int GetInt => a;
    public readonly int GetIntMethod()
    {
        // we could a = 2 if not readonly
        return a;
    }
}
```
You code return true but must false. [See](https://sharplab.io/#v2:C4LglgNgPgAgTARgLACgDOwBOBXAxsAAgGVUBvVAyggB0zADcBDYAUwLADtDGBuCqogApOhagEp+lciiqyCjAgF4CCPjKoBfSQRgBmdlwIBxFsACShxQD55a2XoKYWjACYB7DhACeBwifNcALKmABZuLoIS6lLasgD0cQQA7my4btgQLvJKBHDsAGYEHG6ETq4e3rFUMADsttpaKBpAA) then compare with `private readonly int a;`.

Also the method is all for performance so we shouldn't address all rare cases. We should only reduce false positives in original code.

>  even if we could check it more heuristically like you suggest

I mean new C# 8.0 feature https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/struct#readonly-instance-members

Reply:
Right, but that's a limited case, and doesn't account for (as an example) a method being provided that mutates the private field. We don't have a way to ensure that a settable field is not being set, so it will reduce the false positives more effectively to simply check for settable fields.

I'm not sure I necessarily agree with your reasoning that it's only worth improving in certain ways. If we choose to compare lists of properties and methods as well as fields, in many cases we will end up needing to make _more_ reflection calls than only iterating the fields.

Reply:
>  but that's a limited case

Whole the method is limited case because it is only for value types and most of calls will fast return on first condition `type.IsPrimitive`.
_But it is BINDER code that is very-very sensitive_ and we need strong motivation to change it. Currently no use case exists we should address and no reliable tests and measures too.

Reply:
Right, that's why the current flawed assumptions are concerning.

But if it helps, I can add a case for primitives/enums back in, it most likely would short-circuit things a bit quicker.

Reply:
Added the short-circuiting for enums and primitives back in and tidied things up a little bit.